### ISSUE TITLE
-nan(ind) error with count: 0 for YOLOV3

### ISSUE THREAD

**Issue by renoldhuman** [AlexeyAB/darknet#1454]
I am currently trying to train YOLO to detect a single class within the given data set. Here's a little more information about the dataset. 
    -All of the supplied images are 416x416 jpgs
    -About half of the dataset contains the image being searched for while the other half does not

I have been having an extremely frustrating error where every output for the data set i'm training is -nan(ind) with an object count of 0 this continues to happen even after several thousand iterations.

Here is a screenshot of what I'm talking about

![-nan ind](https://user-images.githubusercontent.com/25161748/44348534-ad8c3080-a468-11e8-8fef-7103e5155945.jpg)

As you can see however my avg loss is not NaN, so according to the documentation...
    "Note: If during training you see nan values for avg (loss) field - then training goes wrong, but if nan is 
    in some other lines - then training goes well."
My training has not gone wrong, but it is not going well.

I have confirmed that all of my annotation files are correct using yolo_mark so that is not a problem, but here is a sample annotation anyway.
  `  0 0.194712 0.329327 0.072115 0.055288`

My first thoughts here are that the object that I am searching for is small, occupying less than 10% of the total image which would explain why the count would be 0 if YOLO cannot find the object because it is so small. This also might explain why everything is NaN, a divide by 0 is occurring somewhere.

I have followed the steps recommended by AlexeyAB for improving object detection but none of them have had any effect on the output
    - I have tried increasing the network size in my cfg file to 608x608
    - I have tried different combinations of batch size and subdivisions, 64/16, 64/8, 128/16, etc...
    - I have ran with `random=1`
    - I have ran with `layers = -1, 11` and `stride = 4` which is supposed to help with small object detection

I have also tried recalculating the anchors for my image many different times and different ways, however I seem to be having another issue here all together in that the `calc_anchors` command gives me an assertion error when I try to run it that looks like this. (Maybe I'll open another issue just for this?)
![kmeans error](https://user-images.githubusercontent.com/25161748/44350220-cac2fe00-a46c-11e8-8795-fda7b3b85461.jpg)

I have managed to get around this error somewhat by using the gen_anchors.py script that is used for YOLOV2 and multiplying the results by 32, but this is a hack and could be fussing up my results idk. _However I get -nan(ind) whether I use the default anchors or the new anchors it doesn't matter._

Here is the command that I'm using to run this training 
`.\darknet detector train .\data\ppoint.data .\yolov3-ppoint-v2.cfg .\darknet53.conv.74`

And here's the copy of the current .cfg file that I'm using for the training
**Note**: The anchors here are the custom ones I got from using the gen_anchors.py script
Any help anyone can give is greatly appreciated as I really am at a loss for what to try next.
```
[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
#batch=128
subdivisions=8
#subdivisions=16
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 2500
policy=steps
steps=1600,1800
scales=.1,.1

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

# Downsample

[convolutional]
batch_normalize=1
filters=64
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=32
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=128
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=256
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=512
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

######################

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear


[yolo]
mask = 6,7,8
anchors = 17.7500,16.7499, 25.6786,16.8571, 24.9999,20.2083, 28.7562,19.6341, 20.8751,27.7500, 27.5501,23.6000, 31.2916,23.1250, 40.0313,22.4375, 44.0968,25.6451
classes=1
num=9
jitter=.3
ignore_thresh = .5
truth_thresh = 1
random=07


[route]
layers = -4

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 61



[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear


[yolo]
mask = 3,4,5
anchors = 17.7500,16.7499, 25.6786,16.8571, 24.9999,20.2083, 28.7562,19.6341, 20.8751,27.7500, 27.5501,23.6000, 31.2916,23.1250, 40.0313,22.4375, 44.0968,25.6451
classes=1
num=9
jitter=.3
ignore_thresh = .5
truth_thresh = 1
random=0



[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 36



[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear


[yolo]
mask = 0,1,2
anchors = 17.7500,16.7499, 25.6786,16.8571, 24.9999,20.2083, 28.7562,19.6341, 20.8751,27.7500, 27.5501,23.6000, 31.2916,23.1250, 40.0313,22.4375, 44.0968,25.6451
classes=1
num=9
jitter=.3
ignore_thresh = .5
truth_thresh = 1
random=0

```

**Comment by AlexeyAB** [comment]
Try to train at least 1000 iterations. Will there be always `count=0`?

> -About half of the dataset contains the image being searched for while the other half does not

Or may be `train.txt` contain only one half of dataset - the images without objects.

Also change `random=07` to `random=0` in your cfg-file.

**Comment by renoldhuman** [comment]
I've run for the full 2500 iterations specified in the cfg at least a couple of times but I will retrain to 1000 iterations now that i've fixed that typo.

The train.txt contains a random sampling of 2/3rds of the full data set that looks something like this

data/ppoint/68213-20180414-1919923.jpg
data/ppoint/68213-20180414-1919923_EMPTY.jpg

where the top filepath is a an example path to an image that contains the object, and the bottom filepath with the "EMPTY" designator is an example path to an image that does not contain the object, so it's definitely a mix of both

Thanks for the help Alexey

**Comment by AlexeyAB** [comment]
Also can you show screenshot of this command?

`./darknet detector calc_anchors .\data\ppoint.data -num_of_clusters 9 -width 416 -height 416`

**Comment by renoldhuman** [comment]
@AlexeyAB 
Here's a small screenshot of the directory containing the pictures along with their labels, is there something I'm missing? All of the photos along with their .txt labels are in this directory (Ignore the !'s, uncommitted Hg files)

![directory](https://user-images.githubusercontent.com/25161748/44362203-45504580-a48e-11e8-89e7-588021e64b88.jpg)

**Comment by AlexeyAB** [comment]
List of images that is specified in the `.\data\ppoint.data`-file doesn't contain filenames of images that have correspond txt-files with non-empty labels.

* Show content of `.\data\ppoint.data`

* Then show content of any txt-file that have objects

* Then get filename of this txt-file and try to find it in the train.txt that specified in the `.\data\ppoint.data`

**Comment by renoldhuman** [comment]
Here's the content of `.\data\ppoint.data`
![ppointdata](https://user-images.githubusercontent.com/25161748/44364146-3c627280-a494-11e8-9788-f3679d61a8d9.jpg)

Here's the content of a txt-file that contains an object in the directory
![labelwobject](https://user-images.githubusercontent.com/25161748/44364187-5603ba00-a494-11e8-94df-5482df031a01.jpg)

And here's the filename of that txt file in the train.txt
![intrain](https://user-images.githubusercontent.com/25161748/44364231-6fa50180-a494-11e8-81df-582be11c838a.jpg)

**Comment by AlexeyAB** [comment]
This is strange.
Can you compress `.\data\ppoint.data`, `.\data\ppoint_train.txt` and all txt-labels files, and attach this archive to your message?

**Comment by renoldhuman** [comment]
Here you go, keep in mind there are more txt labels than there are images in the ppoint_train.txt, as the filepaths to the images are split between ppoint_train.txt and ppoint_test.txt

[ppointyolo.zip](https://github.com/AlexeyAB/darknet/files/2303895/ppointyolo.zip)

**Comment by AlexeyAB** [comment]
@renoldhuman 
I fixed it. Update your code from GitHub.

There are extra spaces at the end of lines in the `ppoint_train.txt` file. I added `trim()` function that removes extra spaces from the end and begining of lines for label path.

**Comment by renoldhuman** [comment]
Amazing thank you so much Alexey

**Comment by ZJJTSL** [comment]
![image](https://user-images.githubusercontent.com/30379945/50479351-12582880-0a11-11e9-8313-6a528aed2dd1.png)
@AlexeyAB  sry to ask my questions. here i followed your command .if it shows any problems? .
